### PR TITLE
Rewrite Buffer.BlockCopy in C#

### LINKED ISSUE #27106: Buffer::BlockCopy may spend too long without GC polling
`Buffer::BlockCopy` is an FCALL that delegates to the native `memmove` which may spend a lot of time there depending on how much is being copied.

If GC needs to sync with user threads at inconvenient time, everything will stop until the `memmove` is done. GC would wait for `memmove`, everything else will wait for GC.
There was an actual scenario reported when GC pauses could take up to a minute due to this.
(dealing with very large streams, potentially swapped out, ...)

We should either "chunk" large copying into smaller pieces with intermittent GC polling, or just move the whole thing to managed code.

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #27106

### CODE REVIEW

File: src/vm/jithelpers.cpp [hunk 1]
```diff
@@ -4092,33 +4091,6 @@ HCIMPL1(Object*, JIT_GetRuntimeMethodStub, CORINFO_METHOD_HANDLE method)
 }
 HCIMPLEND
 
-HCIMPL2(VOID, JIT_GetRuntimeTypeHandle, Object ** destPtr, CORINFO_CLASS_HANDLE type)
```
Comment (jkotas) [inline src/vm/jithelpers.cpp]: Deleted dead/unreachable code that I have run into along the way

**Comment by jkotas** [comment]
Performance results:

Before:

|        Method | numElements |       Mean |
|-------------- |------------ |-----------:|
| CallBlockCopy |          10 |   8.049 ns |
| CallBlockCopy |         100 |  10.297 ns |
| CallBlockCopy |        1000 |  30.870 ns |
| CallBlockCopy |       10000 | 220.712 ns |

After:

|        Method | numElements |       Mean |
|-------------- |------------ |-----------:|
| CallBlockCopy |          10 |   8.862 ns |
| CallBlockCopy |         100 |  11.137 ns |
| CallBlockCopy |        1000 |  41.078 ns |
| CallBlockCopy |       10000 | 227.914 ns |

The managed implementation is slightly slower on average (<5%), but it is GC pause friendly. Achieving the GC pause friendliness in the unmanaged implementation would add more significant overhead.

The larger delta for 1000 elements has to do with tuning of the managed Memcpy implementation that apparently performs worse for this specific block size on my machine.

File: src/vm/typehandle.inl [hunk 2]
```diff
@@ -283,6 +283,53 @@ inline void TypeHandle::ForEachComponentMethodTable(T &callback) const
     }
 }
 
+#ifndef CROSSGEN_COMPILE
+FORCEINLINE OBJECTREF TypeHandle::GetManagedClassObjectFast() const
```
Comment (jkotas) [inline src/vm/typehandle.inl]: This is a general improvement for `typeof(T[])`.

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 3]
```diff
@@ -181,28 +190,38 @@ internal static unsafe bool ObjectHasComponentSize(object obj)
...
             // So in effect this method is the equivalent of
             // return ((MethodTable*)(*obj))->IsStringOrArray();
+            return (int)GetObjectMethodTablePointer(obj)->Flags < 0;
+        }
 
-            Debug.Assert(obj != null);
-            return *(int*)GetObjectMethodTablePointer(obj) < 0;
+        // Subset of src\vm\methodtable.h
+        [StructLayout(LayoutKind.Explicit)]
+        private struct MethodTable
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Baby steps towards https://github.com/dotnet/coreclr/issues/18111

File: src/System.Private.CoreLib/shared/System/Buffer.cs [hunk 4]
```diff
@@ -26,6 +26,54 @@ namespace System
...
+                uSrcLen *= (nuint)src.GetElementSize();
+            }
+
+            nuint uDstLen = uSrcLen;
+            if (src != dst)
+            {
+                uDstLen = (nuint)dst.LongLength;
+                if (dst.GetType() != typeof(byte[]))
+                {
+                    if (!IsPrimitiveTypeArray(dst))
```
Comment (EgorBo) [inline src/System.Private.CoreLib/shared/System/Buffer.cs]: Don't MT flags also contain info about is it primitive or not? `TruePrimitive`

File: src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs [hunk 5]
```diff
@@ -164,12 +171,14 @@ internal static bool IsBitwiseEquatable<T>()
         internal static ref byte GetRawSzArrayData(this Array array) =>
             ref Unsafe.As<RawArrayData>(array).Data;
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static unsafe ref byte GetRawArrayData(this Array array) =>
+            ref Unsafe.AddByteOffset(ref Unsafe.As<RawData>(array).Data, (nuint)GetObjectMethodTablePointer(array)->BaseSize - (nuint)(2 * sizeof(IntPtr)));
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: This works on mono's Array as well?

Comment (stephentoub) [reply src/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.cs]: Oh, nevermind, this isn't shared.

File: src/vm/typehandle.inl [hunk 6]
```diff
@@ -283,6 +283,53 @@ inline void TypeHandle::ForEachComponentMethodTable(T &callback) const
...
+    }
+    CONTRACTL_END;
+
+    OBJECTREF o = NULL;
 
+    if (!IsTypeDesc()) {
+        o = AsMethodTable()->GetManagedClassObjectIfExists();
+    }
+    else
+    {
```
Comment (stephentoub) [inline src/vm/typehandle.inl]: Nit: seems to have an inconsistent style about where the `{` goes, with it being on the same line as the `if`

**Review by stephentoub — APPROVED** [review]

File: src/vm/typehandle.inl [hunk 7]
```diff
@@ -283,6 +283,53 @@ inline void TypeHandle::ForEachComponentMethodTable(T &callback) const
...
+        case ELEMENT_TYPE_PTR:
+            o = dac_cast<PTR_ParamTypeDesc>(AsTypeDesc())->GetManagedClassObjectFast();
+            break;
+
+        case ELEMENT_TYPE_VAR:
+        case ELEMENT_TYPE_MVAR:
+            o = dac_cast<PTR_TypeVarTypeDesc>(AsTypeDesc())->GetManagedClassObjectFast();
+            break;
+
+            // for this release a function pointer is mapped into an IntPtr. This result in a loss of information. Fix next release
```
Comment (VSadov) [inline src/vm/typehandle.inl]: Is the comment still relevant? We return null for FNPTR.

**Review by VSadov — APPROVED** [review]
NICE!

**Comment by benaadams** [comment]
> The managed implementation is slightly slower on average (<5%)

Memmove has scope to be improved; its setup for ymm sizes however it will only ever compile to xmm as it was implemented pre-intrinsics. 

May be worth switching `Block16` for `Vector128` and `Block64` for 2x `Vector256`? However, beyond that; probably not worth chasing down the rabbit hole for 5% as we've been here before several times and its very time consuming.

**Comment by benaadams** [comment]
> Maybe seeing if switching Block16 for Vector128 and Block64 for 2x Vector256 gives an improvement?

Even have a commit for it 😄  https://github.com/benaadams/coreclr/commit/5b90be64652a803f060ed49006cbbe44b2703adf

File: src/System.Private.CoreLib/shared/System/Buffer.cs (hunk 4 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Buffer.cs]: There is no flag that does exactly what this needs today.

**Comment by jkotas** [comment]
> benaadams@5b90be6

That was https://github.com/dotnet/coreclr/pull/25172 . Trying to hand-tune memmove in software is a losing battle. I hope that hardware will add memmove instruction one day that will be superior to hand tuned implementations in every dimension.

**Comment by benaadams** [comment]
> > benaadams/coreclr@5b90be6

> That was #25172 .

😄 so it was; thought it strange I'd not done anything with it!

>  I hope that hardware will add memmove instruction one day that will be superior to hand tuned implementations in every dimension.

Agreed. its such a common operation; I still don't understand why isn't a simple instruction that then can dispatch to the memory or DMA controllers to do the work if the sizes makes sense, rather than have to write a CPU loop in software; and the the cpu can get on other other work or sleep...

**Comment by tannergooding** [comment]
> I hope that hardware will add memmove instruction one day that will be superior to hand tuned implementations in every dimension.

I'm pretty sure this is meant to be the `ERMSB` support and it does basically what you ask. Two problems are that it isn't supported everywhere yet and it has some overhead that makes it undesirable for small loops.
* AMD doesn't support `ERMSB` directly, but it does support block copies for the same instructions. It has a lot of notes on when it is/isn't beneficial to use.
* We could expose hardware intrinsics for `IsGenuineIntel`, `IsAuthenticAmd`, and `ERMSB` to allow this to be correctly handled.

> Trying to hand-tune memmove in software is a losing battle.

I would agree that hand-tuning to have the best perf is likely a losing battle, but many of the rules around copying blocks of memory are well-defined and documented at this point (namely in the respective architecture manuals). It basically comes down to handling sizes less than 128 bytes and then everything else. The split at 128-bytes is defined because that is how much data a `prefetch` will grab.

If we exposed intrinsics for the above, we could just have some code like:
```csharp
if (size < 128)
{
    // small copy
}
else if (Cpuid.IsGenuineIntel && Ermsb.IsSupported)
{
    Ermsb.MoveBytes(src, dst, count);
}
else
{
   // large copy in 64-byte chunks using non-temporal loads/stores
}
```

This should provide overall decent performance and fairly closely match what is recommended by the architecture manuals and done by other memcpy implementations.

**Comment by benaadams** [comment]
Compared to AVX; ERMSB isn't great below 128bytes and its +2% win starts at >= 2048 bytes

image


Also its sounds more sensitive to misalignment than SIMD (+20%)

image

**Comment by benaadams** [comment]
Saying that... I would be interested in a `Cpuid.IsGenuineIntel` test as some of the Ryzen latencies aren't great (e.g. `_pdep_u32` and `_pext_u32`)

**Comment by benaadams** [comment]
Would this be a 3.1 (LTS) candidate?

**Comment by jkotas** [comment]
> Would this be a 3.1 (LTS) candidate?

Unlikely

**Comment by AntonLapounov** [comment]
> slightly slower on average (<5%), but it is GC pause friendly

For 10K elements we fall back to the non-friendly PInvoke: https://github.com/jkotas/coreclr/blob/managed-blockcopy/src/System.Private.CoreLib/shared/System/Buffer.cs#L468-L471
so it seems the actual performance hit is greater than 5%.  Try disabling that fallback.

**Comment by AntonLapounov** [comment]
> The managed implementation is slightly slower on average (<5%), but it is GC pause friendly.

(Corrected the question.) For 10K block size we fall back to a PInvoke, and for 10 to 1K block sizes the performance hit of the managed implementation seems more than 5%.  Is the 1K size indeed somehow exceptional or is that actually evidence that we'd better lower `MemmoveNativeThreshold` to fall back to the PInvoke earlier (at least for your particular machine and the current managed implementation that does not take advantage of intrinsics)?

**Comment by jkotas** [comment]
> 10 to 1K block sizes the performance hit of the managed implementation seems more than 5%

Yeah, you are right. The difference can be up to 20% around the MemmoveNativeThreshold. This is memory copy implementation is used in many other places in managed code. If there is a good reason to look at tuning it again, I would rather do it in a separate change.

> we'd better lower MemmoveNativeThreshold to fall back to the PInvoke earlier 

The current MemmoveNativeThreshold is 2k. The PInvoke overhead is still significant for this buffer size. On my machine, there is actually a performance drop as we cross this threshold and fallback to PInvoke. This would call for making this threshold higher.

File: tests/CoreFX/CoreFX.issues.rsp [hunk 8]
```diff
@@ -99,3 +99,7 @@
 -nomethod System.Numerics.Tests.GenericVectorTests.GetHashCodeInt64
 -nomethod System.Numerics.Tests.GenericVectorTests.GetHashCodeSingle
 -nomethod System.Numerics.Tests.GenericVectorTests.GetHashCodeUInt32
+
+# corefx tests needs to be updated with updated argument name
```
Comment (jkotas) [inline tests/CoreFX/CoreFX.issues.rsp]:
```suggestion
# corefx tests need to be updated with updated argument name
```

**Comment by Robird** [comment]
> > [benaadams/coreclr@5b90be6](https://github.com/benaadams/coreclr/commit/5b90be6)
> 
> That was #25172 . Trying to hand-tune memmove in software is a losing battle. I hope that hardware will add memmove instruction one day that will be superior to hand tuned implementations in every dimension.

I think, acutly there is hardware DMA on broad platforms, rather than CPU register and load/store instructions.

**Pull request merged**